imap_processing/ccsds/excel_to_xtce.py

```diff
@@ -0,0 +1,426 @@
+            parameter_type.attrib["signed"] = "false"
+
+            encoding = Et.SubElement(parameter_type, "xtce:IntegerDataEncoding")
+            encoding.attrib["sizeInBits"] = str(parameter_data["lengthInBits"])
+            encoding.attrib["encoding"] = "unsigned"
+
+    def _create_container_sets(self) -> None:
+        """Create a container set for each packet in the Excel file."""
+        # Iterate over all packets and create Packet SequenceContainers
+        for packet_name, apid in self._packet_mapping.items():
+            # Populate EntryList for packet SequenceContainers
+            # The sheets are sometimes prefixed with P_, so we need to try both options
+            try:
+                packet_df = self.sheets[packet_name]
+            except KeyError:
+                try:
+                    packet_df = self.sheets[f"P_{packet_name}"]
+                except KeyError:
+                    print(
+                        f"Packet definition for {packet_name} "
+                        "not found in the excel file."
+                    )
+                    continue
+
+            # Create Packet SequenceContainer that use the CCSDSPacket SequenceContainer
+            # as the base container
+            science_container = Et.SubElement(
+                self._container_sets, "xtce:SequenceContainer"
+            )
+            science_container.attrib["name"] = packet_name
+
+            # Every container should inherit from the base container, CCSDSPacket
+            base_container = Et.SubElement(science_container, "xtce:BaseContainer")
+            base_container.attrib["containerRef"] = "CCSDSPacket"
+
+            # Add RestrictionCriteria element to use the given APID for comparison
+            restriction_criteria = Et.SubElement(
+                base_container, "xtce:RestrictionCriteria"
+            )
+            comparison = Et.SubElement(restriction_criteria, "xtce:Comparison")
+            comparison.attrib["parameterRef"] = "PKT_APID"
+            comparison.attrib["value"] = str(apid)
+            comparison.attrib["useCalibratedValue"] = "false"
+
+            packet_entry_list = Et.SubElement(science_container, "xtce:EntryList")
+            final_row = len(packet_df) - 1
+            for i, row in packet_df.iterrows():
+                if i < 7:
+                    # Skip first 7 rows as they are the CCSDS header elements
+                    continue
+                if pd.isna(row.get("packetName")):
+                    # This is a poorly formatted row, skip it
+                    continue
+                name = f"{row['packetName']}_{row['mnemonic']}"
+                parameter_ref_entry = Et.SubElement(
+                    packet_entry_list, "xtce:ParameterRefEntry"
+                )
+                parameter_ref_entry.attrib["parameterRef"] = name
+                # Add this parameter to the ParameterSet too
+                self._add_parameter(row, final_two_rows=i >= final_row - 1)
```

Comment:
So this assumes that the last two rows are always dynamic? Is that true for all instruments and HK?

Reply:
Great question. I don't love this myself, but didn't know a better way. Maybe just assume any BYTE value is dynamic is better rather than special casing the two final rows? Let me know your thought on setting everything to be dynamic instead.

We do need the last row as that is most instruments. But Hi has a final fixed-length checksum that comes after their BYTE field, so it is why I added the final two to this.

Reply:
I pushed up a commit to remove the final two and just make everything dynamic for now.